src/Controller/AuthenticationController.php

```diff
@@ -0,0 +1,258 @@
+
+    $request = $this->requestStack->getCurrentRequest();
+    $this->setSessionValue(self::PARAMETERS_NAME, ['query' => $request->query->all()]);
+
+    $provider = new OpenIdConfigurationProvider($providerOptions);
+
+    $authorizationUrl = $provider->getAuthorizationUrl();
+
+    $this->setSessionValue(self::STATE_NAME, $provider->getState());
+
+    return new TrustedRedirectResponse($authorizationUrl);
+  }
+
+  /**
+   * Process OpenID Connect response.
+   */
+  public function process(string $key): Response {
+    $options = $this->getOptions($key);
+
+    $request = $this->requestStack->getCurrentRequest();
+
+    if (!$request->query->has('state') || !$request->query->has('id_token')) {
+      $this->error('Missing state or id_token in response', ['query' => $request->query->all()]);
+      throw new BadRequestHttpException('Missing state or id_token in response');
+    }
+
+    $state = $this->getSessionValue(self::STATE_NAME);
+    if ($state !== $request->query->get('state')) {
+      $this->error('Invalid state', ['state' => $request->query->get('state')]);
+      throw new BadRequestHttpException('Invalid state');
+    }
+
+    // Retrieve id_token and decode it.
+    // @see https://tools.ietf.org/html/rfc7519
+    $idToken = $request->query->get('id_token');
+    [$jose, $payload, $signature] = array_map('base64_decode', explode('.', $idToken));
+    $payload = json_decode($payload, TRUE);
+
+    if (!isset($payload['upn'])) {
+      $this->error('Invalid payload', ['payload' => $payload]);
+      throw new BadRequestHttpException('Invalid payload');
+    }
+
+    try {
+      $user = $this->userHelper->buildUser($payload, $options);
+    }
+    catch (EntityStorageException $exception) {
+      $this->error('Cannot create user', ['exception' => $exception]);
+      throw new BadRequestHttpException('Cannot create user', $exception);
+    }
+    if (!$user) {
+      $this->error('User not created', ['payload' => $payload]);
+      throw new BadRequestHttpException('User not created');
+    }
+
+    user_login_finalize($user);
+
+    $parameters = $this->getSessionValue(self::PARAMETERS_NAME);
+    $location = $parameters['query']['location'] ?? $this->getUrl('<front>');
+
+    return new RedirectResponse($location);
+  }
+
+  /**
+   * Get url from a route.
+   *
+   * Wrapper around Url::fromRoute which does some weird bubbling stuff.
+   *
+   * @param string $name
+   *   The route name.
+   * @param array $parameters
+   *   The route parameters.
+   * @param array $options
+   *   The options.
+   *
+   * @return string
+   *   The url.
+   */
+  private function getUrl(string $name, array $parameters = [], array $options = []) {
+    return Url::fromRoute(
+      $name,
+      $parameters,
+      $options
+    )
+      // Prevent bubble error. @todo documentation!
+      ->toString(TRUE)->getGeneratedUrl();
+  }
+
+  private const STATE_NAME = 'oauth2state';
+  private const PARAMETERS_NAME = 'oauth2params';
```

Comment:
Not really a big thing, but maybe these should be placed in the beginning of the class.

Reply:
Resolved in https://github.com/itk-dev/itkdev_openid_connect_drupal/pull/1/commits/9310fef50b3fe8e6d112193424e22fdbabfab46b